### PR TITLE
Issue #3422694: Send notification for group content that become published

### PR DESCRIPTION
**Pull request by VBaldych**
## Problem
If I create a topic, event, etc. in a group and save it first as a draft (disabling the "published" checkbox), it will not generate a notification for members of the group when it is later updated to be published.

This is also reproducible when using the scheduler extension, no notification is sent when the content inside the group is published later on.

## Solution
- If `entity`, which has relation to `group_content` created unpublished we should not create any activities for message templates that have group_content as trigger entity;
- And vise versa - if `entity` become published (or create as published) we should create activities for message templates that have `group_content` as trigger entity;

## Issue tracker
- https://getopensocial.atlassian.net/browse/PROD-26551
- https://www.drupal.org/project/social/issues/3422694

## Theme issue tracker
n/a

## How to test
- [ ] Login as a SM+
- [ ] Create new VU user
- [ ] Create a group
- [ ] Add a member (created VU user) to the group
- [ ] Create an unpublished Topic inside the group. Save
- [ ] Run cron (/admin/config/system/cron)
- [ ] Go into Mailcatcher through Tugboat Dashboard. You shouldn't see any new emails.
- [ ] Login as a created VU user (in an incognito window). You shouldn't see any new notifications notification about new group content.
- [ ] As an SM+ user edit the created content. Make it published. Save
- [ ] Run cron
- [ ] Go into Mailcatcher. You should see new email
- [ ] As an VU user you also should see a notification about new group content
- [ ] [Optional] Repeat the steps for Event

### Before merge
- [ ] Code/peer review is completed
- [ ] All commit messages are [clear and clean](https://open-social.slite.com/app/docs/DnmermZDIx_0OQ). If applicable a rebase was performed
- [ ] All automated tests are green
- [ ] Functional/manual tests of the acceptance criteria are approved
- [ ] All acceptance criteria were met
- [ ] New features or changes to existing features are covered by tests, either unit (preferably) or behat
- [ ] Update path is tested. New hook_updates should respect update order, right naming convention and consider hook_post_update code
- [ ] Module can be safely uninstalled. Update/implement hook_uninstall and make sure that removed configuration or dependencies are removed/uninstalled
- [ ] This pull request has all required labels (team/type/priority)
- [ ] This pull request has a milestone
- [ ] This pull request has an assignee (if applicable)
- [ ] Any front end changes are tested on all major browsers
- [ ] New UI elements, or changes on UI elements are approved by the design team
- [ ] New features, or feature changes are approved by the product owner

### After merge
- [ ] Code is tested on all branches that it has been cherry-picked
- [ ] Update hook number might need adjustment, make sure they have the correct order
- [ ] The Drupal.org ticket(s) are updated according to this pull request status

## Release notes
Group members get notifications only if group content becomes published

### CODE REVIEW

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3790-dtxhlviziczuq9pjofjyxsyyttchzna4.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/65d4ceca89182614f631e748

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3790-dtxhlviziczuq9pjofjyxsyyttchzna4.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/65d4ceca89182614f631e748

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3790-dtxhlviziczuq9pjofjyxsyyttchzna4.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/65d4ceca89182614f631e748

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3790-dtxhlviziczuq9pjofjyxsyyttchzna4.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/65d4ceca89182614f631e748

**Review by nechai — CHANGES_REQUESTED** [review]

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3790-dtxhlviziczuq9pjofjyxsyyttchzna4.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/65d4ceca89182614f631e748

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3790-dtxhlviziczuq9pjofjyxsyyttchzna4.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/65d4ceca89182614f631e748

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3790-dtxhlviziczuq9pjofjyxsyyttchzna4.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/65d4ceca89182614f631e748

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3790-dtxhlviziczuq9pjofjyxsyyttchzna4.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/65d4ceca89182614f631e748

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3790-dtxhlviziczuq9pjofjyxsyyttchzna4.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/65d4ceca89182614f631e748

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3790-dtxhlviziczuq9pjofjyxsyyttchzna4.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/65d4ceca89182614f631e748

File: modules/custom/activity_basics/activity_basics.module [hunk 1]
```diff
@@ -6,23 +6,34 @@
...
-use Drupal\node\NodeInterface;
 use Drupal\group\Entity\GroupContent;
+use Drupal\group\Entity\GroupContentInterface;
+use Drupal\node\NodeInterface;
 
 /**
  * Implements hook_entity_insert().
  */
 function activity_basics_entity_insert(EntityInterface $entity): void {
+  // Don't trig activity creation for group content entity with
```
[the remark this answers is not in the document]
Comment (VBaldych) [reply modules/custom/activity_basics/activity_basics.module]: Changed!

File: modules/custom/activity_basics/activity_basics.module [hunk 2]
```diff
@@ -38,12 +49,28 @@ function activity_basics_entity_update(EntityInterface $entity): void {
  */
 function activity_basics_entity_presave(EntityInterface $entity): void {
   // Check if the entity become published.
-  if ($entity instanceof EntityPublishedInterface && !$entity->isNew()) {
+  if (
+    $entity instanceof ContentEntityInterface &&
```
[the remark this answers is not in the document]
Comment (VBaldych) [reply modules/custom/activity_basics/activity_basics.module]: Changed!

**Review by nechai — APPROVED** [review]

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3790-dtxhlviziczuq9pjofjyxsyyttchzna4.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/65d4ceca89182614f631e748

File: modules/custom/activity_basics/activity_basics.module [hunk 3]
```diff
@@ -42,8 +53,21 @@ function activity_basics_entity_presave(EntityInterface $entity): void {
     // Check if entity change state from "unpublished" to "published".
     // @phpstan-ignore-next-line
     $original = $entity->original;
+
     if (!$original->isPublished() && $entity->isPublished()) {
       _activity_basics_entity_action($entity, 'publish_entity_action');
+
+      // Load all of the group content for this entity.
+      // Check if related entity become published then trig activity creation.
+      /** @var \Drupal\group\Entity\GroupContentInterface[] $group_contents */
```
Comment (ribel) [inline modules/custom/activity_basics/activity_basics.module:62]: @VBaldych there is PHPStan [warning](https://github.com/goalgorilla/open_social/actions/runs/8064696520/job/22029179024?pr=3790) about this line:
```Variable $group_contents in PHPDoc tag @var does not exist.```

Do you really need this `var` there?
I think issue can be fixed simply by removing this line.

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3790-dtxhlviziczuq9pjofjyxsyyttchzna4.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/65d4ceca89182614f631e748

File: modules/custom/activity_basics/activity_basics.module (hunk 3 above)
Comment (VBaldych) [reply modules/custom/activity_basics/activity_basics.module]: Done

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3790-dtxhlviziczuq9pjofjyxsyyttchzna4.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/65d4ceca89182614f631e748

**Comment by open-social-tugboat** [comment]
[Tugboat](https://www.tugboatqa.com) has finished building the preview for this pull request!

Link:
- https://pr3790-dtxhlviziczuq9pjofjyxsyyttchzna4.tugboatqa.com


Dashboard:
- https://dashboard.tugboatqa.com/65d4ceca89182614f631e748

**Comment by ribel** [comment]
Changes were tested and works as expected.
image

**Comment by ribel** [comment]
🍒 picked to to the 12.1.x, 12.2.x and 12.3.x

**Pull request merged**